Fixes #5931 - SslConnection should implement getBytesIn()/getBytesOut().

sbordet opened the pull request:
Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5931 - SslConnection should implement getBytesIn()/getBytesOut().

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5931 - SslConnection should implement getBytesIn()/getBytesOut().

Make sure the connection is closed before the assertions.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5931 - SslConnection should implement getBytesIn()/getBytesOut().

Fixed CountDownLatch count.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #5931 - SslConnection should implement getBytesIn()/getBytesOut().

Fixed CountDownLatch count again.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-io/src/main/java/org/eclipse/jetty/io/ssl/SslConnection.java:
```diff
@@ -104,7 +105,10 @@
 
+    private final AtomicReference<HandshakeState> _handshake = new AtomicReference<>(HandshakeState.INITIAL);
     private final List<SslHandshakeListener> handshakeListeners = new ArrayList<>();
+    private final AtomicLong _bytesIn = new AtomicLong();
```
We have been changing to LongAdder in other stats classes, so perhaps we should use here as well?

sbordet replied:
I think `LongAdder` is more suited when it's contended, but here is not so I went for `AtomicLong`.

gregw reviewed (dismissed):
Just a niggle regarding the type of the counter, otherwise OK.

Although part of me really hates counting stats all the time, but I think the complexity of optionally counting them is not worth it.

lachlan-roberts commented:
@sbordet is there any issue that existing usages of `ConnectionStatistics` will be now doubly counting bytes from both `HttpConnection` and `SslConnection`. Someone already using `ConnectionStatistics` to monitor their connections won't be expecting this sudden increase.

gregw commented:
@lachlan-roberts will the double counting happen?   Doesn't `ConnectionStatistics` just see the outermost connection?

lachlan-roberts commented on jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientTLSTest.java:
```diff
@@ -1003,4 +1006,49 @@ public void testForcedNonDomainSNI() throws Exception
+    public void testBytesInBytesOut() throws Exception
+    {
+        // Two connections will be closed: SslConnection and HttpConnection.
+        // Two on the server, two on the client.
```
@gregw if you look at the test here it seems to indicate that both `SslConnection` and `HttpConnection` are recorded by `ConnectionStatistics`.

`ConnectionStatistics` works as a `Connection.Listener` so it would apply to `SslConnection` as well.

gregw replied:
Hmmm is this different on the Server-side?  As I don't see how a Connection.Listener would see any connections wrapped by SslConnection?

lachlan-roberts replied:
The `Connection.Listener` is added as a bean to all connectors on the server, and any `Connection.Listener` beans on the Connector are added to each new connection that is created. So the listener will automatically be added to the `SslConnection`.

gregw replied:
Yeah, so the listener is on SslConnection. Is it also inherited by the Wrapped HttpConnection? If so, then i agree the double counting is wrong.

sbordet committed:
Fixes #5931 - SslConnection should implement getBytesIn()/getBytesOut().

Updated ConnectionStatistics to report both the stats of all connections,
and the stats grouped by connection class.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
@sbordet you forgot to request re-review! I've been waiting for y our checkin!

gregw commented on jetty-io/src/main/java/org/eclipse/jetty/io/ConnectionStatistics.java:
```diff
@@ -78,159 +65,312 @@ public void onOpened(Connection connection)
-
-        _connections.increment();
+        _stats.incrementCount();
+        _statsMap.computeIfAbsent(connection.getClass().getName(), Stats::new).incrementCount();
```
perhaps delegate the computeIfAbsent to a protected method, so that if somebody really wanted to pick and choose which types get counted then they could override that method and only return a Stats if they wanted that specific connection counted and it might be on a different criteria than className.

gregw commented on jetty-io/src/main/java/org/eclipse/jetty/io/ConnectionStatistics.java:
```diff
@@ -78,159 +65,312 @@ public void onOpened(Connection connection)
         if (!isStarted())
             return;
+        onClosed(_stats, connection);
+        Stats stats = _statsMap.get(connection.getClass().getName());
```
Ditto. Put this get call into a protected method so selection can be done on criteria other than class name.

gregw reviewed (dismissed):
Other than a niggle for extensibility, this LGTM

sbordet committed:
Fixes #5931 - SslConnection should implement getBytesIn()/getBytesOut().

Updates after review.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
